syntax = "proto3";
package serial;
import "scripts.proto";

message Scene{
  repeated GameObject game_object = 1;
  Camera camera = 2;
}
message Camera{}

message GameObject{
  oneof game_object_instance {
    BasicGameObject basic_game_object = 1;
    // subclasses of game object
  }
}

message Script {
  oneof script_instance {
    PlayerAnimation player_animation = 1;
    ExampleCameraScript example_camera_script = 2;
    ExampleClockedScript example_clocked_script = 3;
    ReloadSceneScript reload_scene_script = 4;
    CameraScript camera_script = 5;
    ButtonScript button_script = 6;
    ClockedScript clocked_script = 7;
    ExampleScript example_script = 8;
    StraightWeaponScript straight_weapon_script = 9;
    WeaponScript weapon_script = 10;
    BulletScript bullet_script = 11;
    NetworkPlayerScript network_player_script = 12;
    CursorAnimation cursor_animation = 13;
  }
  uint32 go_id = 14;
}

//message BasicScript{}
//message ExampleCameraScript{}
//message ExampleClockedScript{}
//message ExampleScript{}
//message CameraScript{}
//message ClockedScript{}
//message ReloadSceneScript{}

message BasicGameObject{
  repeated GameObject children_objects = 1;
  repeated Script children_scripts = 2;
  repeated string tags = 3;
  Sprite sprite = 4;
  bool visible = 5;
  //TODO Need to be refactored since now it connect objects only to main window
  Box2DType box2d_type = 6;
  RenderLayer render_layer = 7;
  uint32 id = 8;
}

message Sprite{
  message Scale{
    float scale_x = 1;
    float scale_y = 2;
  }
  string texture_path = 1;
  Vector2f position = 2;
  Scale scale = 5;
}
enum RenderLayer {
  BACKGROUND = 0;
  BACKGROUND1 = 1;
  BACKGROUND2 = 2;
  BACKGROUND3 = 3;
  FOREGROUND = 4;
  FOREGROUND1 = 5;
  FOREGROUND2 = 6;
  FOREGROUND3 = 7;
  ONTOP = 8;
};

enum Box2DType{
  STATIC = 0;
  DYNAMIC = 1;
}